renter/rpc.go

```diff
@@ -183,34 +217,151 @@ func (s *Session) RegisterSettings(payment PaymentMethod) (settings rhp.HostSett
 }
 
 // ScanSettings returns the current settings for the host.
-func (s *Session) ScanSettings() (rhp.HostSettings, error) {
-	stream, err := s.session.DialStream()
-	if err != nil {
-		return rhp.HostSettings{}, fmt.Errorf("failed to open stream: %w", err)
-	}
-	defer stream.Close()
-	stream.SetDeadline(time.Now().Add(30 * time.Second))
-
-	if err := rpc.WriteRequest(stream, rhp.RPCSettingsID, nil); err != nil {
-		return rhp.HostSettings{}, fmt.Errorf("failed to write settings request: %w", err)
-	}
-
+func (s *Session) ScanSettings(ctx context.Context) (rhp.HostSettings, error) {
 	var resp rhp.RPCSettingsResponse
-	if err := rpc.ReadResponse(stream, &resp); err != nil {
+	if err := s.call(ctx, rhp.RPCSettingsID, nil, &resp); err != nil {
 		return rhp.HostSettings{}, fmt.Errorf("failed to read settings response: %w", err)
 	}
-
 	var settings rhp.HostSettings
 	if err := json.Unmarshal(resp.Settings, &settings); err != nil {
 		return rhp.HostSettings{}, fmt.Errorf("failed to decode settings: %w", err)
 	}
+	s.settings = settings
 	return settings, nil
 }
 
+// Lock calls the Lock RPC, locking the supplied contract and synchronizing its
+// state with the host's most recent revision. Where applicable, future RPC
+// calls will modify the the locked contract, ending when either the Unlock RPC
+// is called or the Session is closed.
+//
+// Lock returns ErrContractFinalized if the contract can no longer be revised.
+// The contract will still be available via the LockedContract method, but
+// invoking other RPCs may result in errors or panics.
+func (s *Session) Lock(ctx context.Context, id types.ElementID, key types.PrivateKey) (err error) {
+	req := &rhp.RPCLockRequest{
+		ContractID: id,
+		Signature:  s.session.SignChallenge(key),
+	}
+	if deadline, ok := ctx.Deadline(); ok {
+		req.Timeout = uint64(time.Until(deadline).Milliseconds())
+	}
+	var resp rhp.RPCLockResponse
+	if err := s.call(ctx, rhp.RPCLockID, req, &resp); err != nil {
+		return err
+	}
+	s.session.SetChallenge(resp.NewChallenge)
+
+	// verify claimed revision
+	if err := rhp.ValidateContractSignatures(s.cm.TipContext(), resp.Revision); err != nil {
+		return fmt.Errorf("host returned invalid revision: %w", err)
+	}
+	if !resp.Acquired {
+		return ErrContractLocked
+	}
+	s.contract = rhp.Contract{
+		ID:       id,
+		Revision: resp.Revision,
+	}
+	s.renterKey = key
+	if !s.canRevise() {
+		return ErrContractFinalized
+	}
+	return nil
+}
+
+// Read calls the Read RPC, writing the requested sections of sector data to w.
+// Merkle proofs are always requested.
+//
+// Note that sector data is streamed to w before it has been validated. Callers
+// MUST check the returned error, and discard any data written to w if the error
+// is non-nil. Failure to do so may allow an attacker to inject malicious data.
+func (s *Session) Read(ctx context.Context, w io.Writer, sections []rhp.RPCReadRequestSection) (err error) {
```

Comment:
Does a separate `Read` or a session-wide `Lock` still make sense?

Reply:
perhaps not, but I'm resisting the urge to make any significant changes to the RHP for now. We'll probably overhaul a lot of stuff as a result of the upcoming summit.